File: core/src/test/java/com/avast/continuity/javaapi/ContinuityTest.java
package com.avast.continuity.javaapi;

import com.avast.continuity.ContinuityContextThreadNamer;
import com.avast.continuity.ThreadNamer;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.*;

import static org.junit.Assert.assertEquals;

public class ContinuityTest {

    private Logger logger = LoggerFactory.getLogger(this.getClass());

    @Test
    public void testExecutorTraceId() throws InterruptedException {
        ThreadNamer threadNamer = ContinuityContextThreadNamer.prefix("traceId");
        Executor original = Executors.newCachedThreadPool();
        Executor wrapped = Continuity.wrapExecutor(original, threadNamer);

        testExecuteTraceId(threadNamer, original, wrapped);
    }

    @Test
    public void testExecutorServiceTraceId() throws InterruptedException {
        ThreadNamer threadNamer = ContinuityContextThreadNamer.prefix("traceId");
        ExecutorService original = Executors.newCachedThreadPool();
        ExecutorService wrapped = Continuity.wrapExecutorService(original, threadNamer);

        testExecuteTraceId(threadNamer, original, wrapped);
    }

    @Test
    public void testScheduledExecutorServiceTraceId() throws InterruptedException {
        ThreadNamer threadNamer = ContinuityContextThreadNamer.prefix("traceId");
        ScheduledExecutorService original = Executors.newSingleThreadScheduledExecutor();
        ScheduledExecutorService wrapped = Continuity.wrapScheduledExecutorService(original, threadNamer);

        testExecuteTraceId(threadNamer, original, wrapped);
        testScheduleTraceId(threadNamer, original, wrapped);
    }

    private void testExecuteTraceId(ThreadNamer threadNamer, Executor original, Executor wrapped) throws InterruptedException {
        CountDownLatch latch = new CountDownLatch(3);

        String traceId1 = "id1";
        Continuity.withContext(map("traceId", traceId1), threadNamer, () -> {
            wrapped.execute(() -> {
                assertEquals(Continuity.getFromContext("traceId"), Optional.of(traceId1));
                logger.info("first");
            });
            latch.countDown();
            return null;
        });

        String traceId2 = "id2";
        Continuity.withContext(map("traceId", traceId2), threadNamer, () -> {
            original.execute(() -> {
                assertEquals(Continuity.getFromContext("traceId"), Optional.empty());
                logger.info("second");
            });
            latch.countDown();
            return null;
        });

        String traceId3 = "id3";
        Continuity.withContext(map("traceId", traceId3), threadNamer, () -> {
            wrapped.execute(() -> {
                assertEquals(Continuity.getFromContext("traceId"), Optional.of(traceId3));
                logger.info("third");
            });
            latch.countDown();
            return null;
        });

        latch.await();
        Thread.sleep(500); // wait for the logger to write out its buffer
    }

    private void testScheduleTraceId(ThreadNamer threadNamer, ScheduledExecutorService original, ScheduledExecutorService wrapped) throws InterruptedException {
        CountDownLatch latch = new CountDownLatch(3);

        String traceId1 = "id1";
        Continuity.withContext(map("traceId", traceId1), threadNamer, () -> {
            wrapped.schedule(() -> {
                assertEquals(Continuity.getFromContext("traceId"), Optional.of(traceId1));
                logger.info("first");
            }, 10, TimeUnit.MILLISECONDS);
            latch.countDown();
            return null;
        });

        String traceId2 = "id2";
        Continuity.withContext(map("traceId", traceId2), threadNamer, () -> {
            original.schedule(() -> {
                assertEquals(Continuity.getFromContext("traceId"), Optional.empty());
                logger.info("second");
            }, 10, TimeUnit.MILLISECONDS);
            latch.countDown();
            return null;
        });

        String traceId3 = "id3";
        Continuity.withContext(map("traceId", traceId3), threadNamer, () -> {
            wrapped.schedule(() -> {
                assertEquals(Continuity.getFromContext("traceId"), Optional.of(traceId3));
                logger.info("third");
            }, 10, TimeUnit.MILLISECONDS);
            latch.countDown();
            return null;
        });

        latch.await();
        Thread.sleep(500); // wait for the logger to write out its buffer
    }

    @Test
    public void testExecutorFutures() throws Exception {
        ContinuityContextThreadNamer threadNamer = ContinuityContextThreadNamer.prefix("traceId");
        Executor wrapped = Continuity.wrapExecutor(Executors.newCachedThreadPool(), threadNamer);

        testFutures(threadNamer, wrapped);
    }

    @Test
    public void testExecutorServiceFutures() throws Exception {
        ContinuityContextThreadNamer threadNamer = ContinuityContextThreadNamer.prefix("traceId");
        ExecutorService wrapped = Continuity.wrapExecutorService(Executors.newCachedThreadPool(), threadNamer);

        testFutures(threadNamer, wrapped);
    }

    @Test
    public void testScheduledExecutorServiceFutures() throws Exception {
        ContinuityContextThreadNamer threadNamer = ContinuityContextThreadNamer.prefix("traceId");
        ScheduledExecutorService wrapped = Continuity.wrapScheduledExecutorService(Executors.newSingleThreadScheduledExecutor(), threadNamer);

        testFutures(threadNamer, wrapped);
    }

    private void testFutures(ThreadNamer threadNamer, Executor wrapped) throws Exception {
        CompletableFuture<Integer> f = Continuity.withContext(map("traceId", "value"), threadNamer, () -> {
            CompletableFuture<Integer> x = new CompletableFuture<>();
            x.whenCompleteAsync((a, b) -> {
                assertEquals("value", Continuity.getFromContext("traceId"));
                logger.info("inside completablefuture " + Continuity.getFromContext("traceId"));
            }, wrapped);
            x.complete(1);
            return x;
        });

        f.get();
        Thread.sleep(500); // wait for the logger to write out its buffer
    }

    private Map<String, String> map(String key, String value) {
        LinkedHashMap<String, String> map = new LinkedHashMap<>();
        map.put(key, value);
        return map;
    }

}

File: core-scala/src/test/scala/com/avast/continuity/ContinuityTest.scala
package com.avast.continuity

import java.util.concurrent.{Executor, Executors}

import com.avast.utils2.concurrent.ConfigurableThreadFactory
import com.avast.utils2.concurrent.ConfigurableThreadFactory.IndexingNamingStrategy
import org.scalatest.FunSuite
import org.scalatest.concurrent.Waiters
import org.scalatest.time.{Seconds, Span}
import org.slf4j.LoggerFactory

import scala.concurrent.duration.Duration
import scala.concurrent.{Await, ExecutionContext, Future}

class ContinuityTest extends FunSuite with Waiters {

  /** See the log output to check that traceId is present. */
  private val logger = LoggerFactory.getLogger(this.getClass)

  test("traceId") {
    testTraceId(Executors.newSingleThreadExecutor)
    testTraceId(ExecutionContext.global)
  }

  private def testTraceId(pool: Executor): Unit = {
    val waiter = new Waiter

    implicit val threadNamer = ContinuityContextThreadNamer.prefix("traceId")
    val wrappedPool = Continuity.wrapExecutor(pool)

    val traceId1 = "id1"
    Continuity.withContext("traceId" -> traceId1) {
      wrappedPool.execute(new Runnable {
        override def run(): Unit = {
          assert(Continuity.getFromContext("traceId") === Some(traceId1))
          logger.info("first")
          waiter.dismiss()
        }
      })
    }

    val traceId2 = "id2"
    Continuity.withContext("traceId" -> traceId2) {
      pool.execute(new Runnable {
        override def run(): Unit = {
          assert(Continuity.getFromContext("traceId") === None)
          logger.info("second")
          waiter.dismiss()
        }
      })
    }

    val traceId3 = "id3"
    Continuity.withContext("traceId" -> traceId3) {
      wrappedPool.execute(new Runnable {
        override def run(): Unit = {
          assert(Continuity.getFromContext("traceId") === Some(traceId3))
          logger.info("third")
          waiter.dismiss()
        }
      })
    }

    waiter.await(dismissals(3))
  }

  test("futures") {
    implicit val ec = Continuity.wrapExecutionContext(ExecutionContext.global)
    val f = Continuity.withContext("traceId" -> "value") {
      Future {
        assert(Continuity.getFromContext("traceId") === Some("value"))
        logger.info(s"creating future ${ Continuity.getFromContext("traceId") }")
        1
      }.map { i =>
        assert(Continuity.getFromContext("traceId") === Some("value"))
        logger.info(s"mapping future ${ Continuity.getFromContext("traceId") }")
        i + 1
      }
    }

    assert(Await.result(f, Duration("1s")) === 2)
  }

  test("double wrapping") {
    val waiter = new Waiter

    implicit val namer = ContinuityContextThreadNamer.prefix("traceId")
    val pool = ExecutionContext.fromExecutor(
      Executors.newFixedThreadPool(1, ConfigurableThreadFactory.builder.withNamingStrategy(new IndexingNamingStrategy("thread-%d")).build))
    val target = Continuity.wrapExecutionContext(Continuity.wrapExecutionContext(pool))

    val traceId1 = "id1"
    Continuity.withContext("traceId" -> traceId1) {
      target.execute(new Runnable {
        override def run(): Unit = {
          println("running")
          assert(Continuity.getFromContext("traceId") === Some(traceId1))
          assert(Thread.currentThread.getName.startsWith("id1-thread-0"))
          waiter.dismiss()
        }
      })
    }

    waiter.await(timeout(Span(5, Seconds)), dismissals(1))
  }

}
